### PR TITLE
Add functional data section

### PR DESCRIPTION
**Pull request by pfistfl**
This is a new PR for the functional data section, as I could not properly rebase the original branch.
#101 contains some of the discussion and the original pr

### CODE REVIEW

**Comment by pfistfl** [comment]
@SteveBronder 
This does still not pass. Do you have any idea?

**Comment by pfistfl** [comment]
Maybe @jakob-r  can help? I have no idea why I keep getting the missing kknn learner error.

The error is
```
Knitting file 'nested_resampling.Rmd' ...
Quitting from lines 312-338 () 
Error in requirePackages(package, why = stri_paste("learner", id, sep = " "),  : 
  For learner classif.kknn please install the following packages: kknn
Calls: lapply ... addClasses -> makeRLearnerInternal -> requirePackages -> stopf
Execution halted
 ```

**Comment by pat-s** [comment]
@pfistfl I face similar problems locally a lot when running `./build`. Its most often solved by explictly adding a `library()` call into the problematic code chunk.

For travis, this does of course not apply and currently I think its random travis behaviour.

**Comment by jakob-r** [comment]
So the error just occurs on cran randomly but locally everything works fine?

**Comment by pat-s** [comment]
@jakob-r The error occured locally for me while everything was fine on travis.

Its a strange one because the package that is said to be not installed is definitely installed. 
If I added an explicit `libray()` call into the offending code chunk, everything was working fine. 
Maybe this will also help in this case..

**Comment by SteveBronder** [comment]
Maybe this is a `maximum number of DLLs reached` error in disguise? Can you try setting this environment variable

`R_MAX_NUM_DLLS=300`

and re-running. If it compiles on your local then it's not this issue (probably)

**Comment by pfistfl** [comment]
@SteveBronder 
> Maybe this is a maximum number of DLLs reached

I think this must be the case.
Reasons: 
- The error is returned for tutorial sections I have not touched at all 
- The offending package was **kknn** before, and now it is **earth** (changed with new merged pr's)
- I got the same error locally today, and setting `R_MAX_NUM_DLLS=300` solved the issue.

**Comment by pfistfl** [comment]
The functional data section seems to load ~14 dlls. This puts the number of .dlls to over 100.
I added a call to `length(getLoadedDLLs())` to `build`.
See: [out.txt](https://github.com/mlr-org/mlr-tutorial/files/1609577/out.txt)

**Comment by SteveBronder** [comment]
Yeah in that case you should be able to set the above environment variable on travis and everything will compile fine

**Comment by pfistfl** [comment]
Thanks a lot, I did not know it would be that easy. Hope it passes now.

**Comment by SteveBronder** [comment]
Yep! I brought this issue up to R devel a while ago and one guy was nice enough to have make a very clean / simple patch.

If you want another set of eyes to review it before you merge I should have time this week

**Comment by pfistfl** [comment]
Yeah that'd be awesome. I think some people already reviewed the old PR, but reviewing can not hurt. 
Other than that I think we are finally ready to merge.

File: src/functional_data.Rmd
Comment (SteveBronder) [inline src/functional_data.Rmd]: I would add a `library(mlr)` to the top of this so that `makeFunctionalData()` works :)

File: src/functional_data.Rmd
Comment (SteveBronder) [inline src/functional_data.Rmd]: >  Either create a new task from an existing task

or?

File: src/functional_data.Rmd
Comment (SteveBronder) [inline src/functional_data.Rmd]: Can I get both the Fourier and the fpca scores for UVVIS?

File: src/functional_data.Rmd
Comment (SteveBronder) [inline src/functional_data.Rmd]: It could be nice for the user to have the code here to run the `wrapped.lrn` so they can just copy/paste the code and get the end results

File: src/functional_data.Rmd
Comment (SteveBronder) [inline src/functional_data.Rmd]: For newbies, it may be nice to have links to what is the amplitude and phase part of the complex fourier coefficients.

File: src/functional_data.Rmd
Comment (SteveBronder) [inline src/functional_data.Rmd]: I often find it's nice to have a one or two line 'splainer' to say something like, "Here is the jist of what a wavelet is" before you send them off to other docs. You may want to do this above as well

**Review by SteveBronder — COMMENTED** [review]
Just a few odds and ends that you can choose to change at your discretion. Overall I'm very excited to sit down and play with this!

File: src/functional_data.Rmd
Comment (pfistfl) [reply src/functional_data.Rmd]: Yes, trying to make that clearer.

**Comment by pfistfl** [comment]
Thanks a lot, I really found your suggestions helpful.
I had a hard time describing the wavelets and fourier transform in a few lines (heck, I could not even find a short, concise explanation on the internet). 
I added links to tutorials instead, hope that helps.

**Comment by pfistfl** [comment]
@jakob-r @schiffner

I had to change the Travis config.
Pending your ok, I think we are ready for merge

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Could you please set the aspect ratio (fig.asp) instead of height and width? That works better with the pdf version.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: If you just want to load the data, could you please use `data(fuelSubset, package = "FDboost")`?

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Question mark is missing.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: typo: following

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: analyzing

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: I think the file extension in the link should be md, also several more times below.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Typo in link: Should be `[learner](&Learner)`, occurs also several more times below.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: - I think file extension in links should be md
- Looks like a missing blank before the 3rd sentence in l. 37

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Typo: Should be either matrices or matrixes. Also several more times below.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: I think there is a word missing in the first sentence: Should be "... vectors of length 231 and 129 AND indicate ..."

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: - Typo: functional
- Please always start comments with `##`

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Should be "Feature extraction"

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Fourier should be capitalized, also several times below.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Fourier

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Typo: function

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Typo: information

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Link should probably be `[fuelSubset](&fuelSubset.task)`.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: The first `<` is unnecessary.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Please mention the name of the function you are talking about and link to it.
Please use apostrophes for formatting argument names (`pars`, `filter`).

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Please use apostrophes to format argument names (`trafo.coeff`)

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Which section do you mean? If you want to refer to the tutorial page about wrappers please use a link.

File: src/functional_data.Rmd
Comment (schiffner) [inline src/functional_data.Rmd]: Please link to the rpart help page.

**Review by schiffner — CHANGES_REQUESTED** [review]
Thanks very much for writing this!
I left some comments regarding formal stuff (typos, links, formatting).

Julia

**Comment by pfistfl** [comment]
Thanks a lot!
I should have checked the links!

**Comment by pfistfl** [comment]
So basically the builds now break because of travis's time limit.
We had build times of around 49 mins before, so this is to be expected I guess.

Is there anything I can do? Set up travis build stages?

**Comment by SteveBronder** [comment]
If you are planning to set up build stages I would recommend looking over @pat-s work in #115

**Comment by SteveBronder** [comment]
Besides using s3 for build stages I'm not sure how to get around the timer for us. That alt is to help/wait for building in pkgdown within mlr [#2123](https://github.com/mlr-org/mlr/pull/2123) which will probably save a lot of headache in the long run

**Comment by pat-s** [comment]
> If you are planning to set up build stages I would recommend looking over @pat-s work in #115 

I have no plans continuing here in the near future; focusing on the pkgdown approach.
There we have a reduced build time currently ([42 min](https://travis-ci.org/pat-s/mlr)) and adding a new section should work there.

I would not put work in the integration into the mkdocs version here if we switch within the next weeks. 

But maybe @jakob-r has a different opinion on this?

**Comment by pat-s** [comment]
hi @pfistfl,

would you mind integrating your section in the [pkgdown PR](https://github.com/mlr-org/mlr/pull/2123)? You should have rights to add commits. 

See https://github.com/pat-s/mlr#mlr-tutorial for a guide of the new approach.

**Comment by pfistfl** [comment]
@pat-s 
Hi, perfect, this sounds really good. 
I was waiting for that!

**Comment by pfistfl** [comment]
I am on it, I will create a PR into ```pat-s:master``` later. 

One thing, I saw is that ```rpkgdown::serve_site()``` does not exist for me.

**Comment by pat-s** [comment]
> One thing, I saw is that rpkgdown::serve_site() does not exist for me.

Oh, is that a mistake in the doc? It should of course be `pkgdown::build_site`

**Comment by pfistfl** [comment]
Hey, @pat-s, I pushed into ```pat-s:mlr/master``` because I could not create branches. 
Will you take it from there? 
If you need anything from me, let me know!

**Comment by pat-s** [comment]
Thanks, looks good. I just did some [minor changes](https://github.com/pat-s/mlr/commit/9de4b93184ff8e49bdb5a5134fad050d578d6d90):

* Start a new line for every sentence
* Exact one empty line before and after a code chunk
* One package reference was still in the mkdocs style
* The list around "regression, classif" was indented and therefore misaligned. Not sure if its completely fixed now, please check again after the site is deployed
* add to pdf version

I would say you can then finally close this PR :smile: 

EDIT: As long as everybody else is fine with the content. I did not check in detail ;)

**Comment by pat-s** [comment]
I appears to be that section https://pat-s.github.io/mlr/articles/tutorial/devel/functional_data.html#constructing-a-learner needs the `mxnet` package installed to properly display the code output. 
However, the `mxnet` package is not available on CRAN. 
Was this ever an issue in `mlr-tutorial`?
If not, you might wanna try to integrate this into the travis build @pfistfl 
The build process is not completely trivial though: http://mxnet.incubator.apache.org/install/index.html

Furthermore, could you maybe try to remove the single bullet points in this section and see if that improves readability? They look a bit misplaced and are probably not really needed due to the code breaks.

**Comment by pfistfl** [comment]
Hey @pat-s 

Thank you very much!

I will iterate again.

I think the mxnet learners have been added very recently, and they are not connected to my tutorial.

I would just not show the warning for now. Are you ok with that?

Other than that, this should do in order to install mxnet:

```r 
  cran <- getOption("repos")
  cran["dmlc"] <- "https://s3-us-west-2.amazonaws.com/apache-mxnet/R/CRAN/"
  options(repos = cran)
  install.packages("mxnet")
```

**Comment by pfistfl** [comment]
Looks good to me now!

Closing here I guess

**Pull request closed without merging**
